### PR TITLE
feat: kafka bindings with support for scalar parameter values

### PR DESCRIPTION
**Pull request by vearutop**
**Description**

Based on [slack discussion](https://asyncapi.slack.com/archives/C34F2JV0U/p1663919937776929) this PR extends `kafka`  bindings to allow scalar values to optimize for a common case when key/clientId/groupId can be defined as constant.

### CODE REVIEW

File: kafka/README.md [hunk 1]
```diff
@@ -35,8 +35,8 @@ This object contains information about the operation representation in Kafka.
 
 Field Name | Type | Description
 ---|:---:|---
-<a name="operationBindingObjectGroupId"></a>`groupId` | [Schema Object][schemaObject] | Id of the consumer group.
-<a name="operationBindingObjectClientId"></a>`clientId` | [Schema Object][schemaObject] | Id of the consumer inside a consumer group.
+<a name="operationBindingObjectGroupId"></a>`groupId` | [Schema Object][schemaObject] \| string | Id of the consumer group.
+<a name="operationBindingObjectClientId"></a>`clientId` | [Schema Object][schemaObject] \| string | Id of the consumer inside a consumer group.
```
Comment (fmvilas) [inline kafka/README.md:39]:
```suggestion
<a name="operationBindingObjectGroupId"></a>`groupId` | string \| [Schema Object][schemaObject] | Id of the consumer group or a [Schema Object][schemaObject] that validates it.
<a name="operationBindingObjectClientId"></a>`clientId` | [string \| [Schema Object][schemaObject] | Id of the consumer inside a consumer group or a [Schema Object][schemaObject] that validates it.
```

File: kafka/README.md [hunk 2]
```diff
@@ -69,7 +69,7 @@ This object contains information about the message representation in Kafka.
 
 Field Name | Type | Description
 ---|:---:|---
-<a name="messageBindingObjectKey"></a>`key` | [Schema Object][schemaObject] \| [AVRO Schema Object](https://avro.apache.org/docs/current/spec.html) | The message key. **NOTE**: You can also use the [reference object](https://asyncapi.io/docs/specifications/v2.1.0#referenceObject) way. 
+<a name="messageBindingObjectKey"></a>`key` | [Schema Object][schemaObject] \| [AVRO Schema Object](https://avro.apache.org/docs/current/spec.html) \| string | The message key. **NOTE**: You can also use the [reference object](https://asyncapi.io/docs/specifications/v2.1.0#referenceObject) way.
```
Comment (fmvilas) [inline kafka/README.md:72]:
```suggestion
<a name="messageBindingObjectKey"></a>`key` | string \| [Schema Object][schemaObject] \| [AVRO Schema Object](https://avro.apache.org/docs/current/spec.html) | The message key or an object that validates the key. **NOTE**: You can also use a [Reference Object](https://asyncapi.io/docs/specifications/v2.1.0#referenceObject).
```

**Review by fmvilas — CHANGES_REQUESTED** [review]
Left some comments. Some of the suggestions are not related to your modifications but I thought we could take the opportunity now.

Also, I'm missing that you bump the version of the document at the top. We don't have any automation there yet for bindings so you have to change it manually.

**Review by fmvilas — DISMISSED** [review]
LGTM 👍 

Let's wait for @derberg, @lbroudoux, @KhudaDad414, and @dalelane reviews.

**Comment by derberg** [comment]
as @lbroudoux and @dalelane are codeowners for kafka binding and @KhudaDad414 maintains JSON Schemas, I will not do the review here - already enough 👀 here 😃

**Comment by fmvilas** [comment]
Oh yeah, forgot that Khuda maintains the schemas.

File: kafka/README.md [hunk 3]
```diff
@@ -35,8 +35,8 @@ This object contains information about the operation representation in Kafka.
 
 Field Name | Type | Description
 ---|:---:|---
-<a name="operationBindingObjectGroupId"></a>`groupId` | [Schema Object][schemaObject] | Id of the consumer group.
-<a name="operationBindingObjectClientId"></a>`clientId` | [Schema Object][schemaObject] | Id of the consumer inside a consumer group.
+<a name="operationBindingObjectGroupId"></a>`groupId` | string \| [Schema Object][schemaObject] | Id of the consumer group or a [Schema Object][schemaObject] that validates it.
+<a name="operationBindingObjectClientId"></a>`clientId` | [string \| [Schema Object][schemaObject] | Id of the consumer inside a consumer group or a [Schema Object][schemaObject] that validates it.
```
Comment (KhudaDad414) [inline kafka/README.md:39]:
```suggestion
<a name="operationBindingObjectClientId"></a>`clientId` | string \| [Schema Object][schemaObject] | Id of the consumer inside a consumer group or a [Schema Object][schemaObject] that validates it.
```

**Review by KhudaDad414 — CHANGES_REQUESTED** [review]
LGTM! just a little typo.

**Review by KhudaDad414 — APPROVED** [review]

File: kafka/README.md [hunk 4]
```diff
@@ -69,7 +69,7 @@ This object contains information about the message representation in Kafka.
 
 Field Name | Type | Description
 ---|:---:|---
-<a name="messageBindingObjectKey"></a>`key` | [Schema Object][schemaObject] \| [AVRO Schema Object](https://avro.apache.org/docs/current/spec.html) | The message key. **NOTE**: You can also use the [reference object](https://asyncapi.io/docs/specifications/v2.1.0#referenceObject) way. 
+<a name="messageBindingObjectKey"></a>`key` | string \| [Schema Object][schemaObject] \| [AVRO Schema Object](https://avro.apache.org/docs/current/spec.html) | The message key or an object that validates the key. **NOTE**: You can also use a [Reference Object](https://asyncapi.io/docs/specifications/v2.1.0#referenceObject).
```
Comment (dalelane) [inline kafka/README.md:72]:
```suggestion
<a name="messageBindingObjectKey"></a>`key` | string \| [Schema Object][schemaObject] \| [AVRO Schema Object](https://avro.apache.org/docs/current/spec.html) | The message key or an object that validates the key. **NOTE**: You can also use a [Reference Object](https://asyncapi.io/docs/specifications/v2.4.0#referenceObject).
```

worth updating this while we're here?

**Review by dalelane — APPROVED** [review]
looks good - I think this will be helpful, thanks! 👍

File: kafka/README.md (hunk 4 above)
Comment (lbroudoux) [reply kafka/README.md:72]: +1 for @dalelane suggestion

File: kafka/json_schemas/message.json [hunk 5]
```diff
@@ -11,13 +11,22 @@
...
     "key": {
-      "$ref": "https://raw.githubusercontent.com/asyncapi/asyncapi-node/v2.7.7/schemas/2.0.0.json#/definitions/schema",
+      "oneOf": [
+        {
+          "description": "Constant value of key.",
+          "type": "string"
+        },
+        {
+          "description": "Schema that describes the allowed values of key.",
+          "$ref": "https://raw.githubusercontent.com/asyncapi/asyncapi-node/v2.7.7/schemas/2.0.0.json#/definitions/schema"
```
Comment (lbroudoux) [inline kafka/json_schemas/message.json:21]: Should be updated to `https://raw.githubusercontent.com/asyncapi/asyncapi-node/v2.14.0/schemas/2.4.0.json#/definitions/schema`

**Review by lbroudoux — CHANGES_REQUESTED** [review]
LGTM. Just update schema references in `message.json` and we're good to go.

**Comment by vearutop** [comment]
Oh, sorry for delay, I see these bindings were recently versioned with 0.3.0 for other changes, should I bump version to 0.3.1 (or to 0.4.0) here?

**Comment by fmvilas** [comment]
@vearutop Yeah, bump it to 0.4.0 as it's actually a new "feature".

**Review by dalelane — DISMISSED** [review]

**Review by fmvilas — APPROVED** [review]
LGTM 👍

**Comment by derberg** [comment]
sorry folks, I wrote I won't review, but actually, today during 3.0 spec meeting it popped up and I read the thread on slack because of it.

can you help me understand why it is better to allow just plain string instead of following schema object:
```
          groupId:
            const: my-group
```

isn't it only a problem of bad example that suggest `enum` while better to have `const`?

**Comment by vearutop** [comment]
Good thing about scalar value is that it has clear intent and is hard to misuse. When I first saw that `groupId` accepts schema I was genuinely confused (by that time I haven't checked examples yet). 

So in my opinion schema covers advanced cases where user would have to look in the docs anyway, while "string" simplifies basic case (constant value). It also works a bit like documentation, suggesting the type of value (schema does not limit itself in any way and lends itself for misuse if user is not careful about what goes into schema).

**Comment by fmvilas** [comment]
Yeah, also a plain string is always a scalar value. A Schema may not be possible to convert down to a scalar value if it doesn't have `const` or an `enum` with a single element array. They serve two different purposes. A producer may want to know which exact value to send. A consumer may want to know how to validate it.

For instance, a publisher may send `key` as `user.signedupV1` and another publisher may set it to `user.signedupV2` but a potential consumer may just want to check that the value is either `user.signedupV1` or `user.signedupV2`, or that it simply matches a specific regex pattern.

The publishers want to have the scalar value they send. The consumer wants to have a Schema object that validates it. Does it make sense?

**Comment by dalelane** [comment]
@lbroudoux do you think that covers everything you saw?

**Comment by derberg** [comment]
yeah, I still think better is:

```yaml
groupId:
    const: my-group
    description: You can always specify more details for the user if you want
```

than:
```yaml
groupId: my-group
```

anyway, I suggest you provide an example for string and probably adjust example with single element `enum` to `const`

**Comment by lbroudoux** [comment]
Still looking good to me. 👍 

Also I think it could be nice to illustrate both cases through different examples like:


```yaml
channels:
  test:
    publish:
      message:
        bindings:
          kafka:
            key: 'user.signedupV1'
```

OR

```yaml
channels:
  test:
    publish:
      message:
        bindings:
          kafka:
            key:
              type: string
              enum: ['user.signedupV1', 'user.signedupV2']
```

IMO this clearly illustrates that you can use this notation shortcut for self-explanatory values but have to switch to schema notation once you handle more complex cases.  In that later cases, additional `description` can make sense to explain the range of possible values.

**Review by lbroudoux — APPROVED** [review]

**Comment by vearutop** [comment]
Hi, is there anything I can do to improve this PR?

**Comment by fmvilas** [comment]
@vearutop Yeah, I think it's just a matter of adding examples as @derberg and @lbroudoux suggested.

**Pull request closed without merging**